pkg/channel/jetstream/dispatcher/reconciler.go

```diff
@@ -0,0 +1,190 @@
+
+func (r *Reconciler) ReconcileKind(ctx context.Context, nc *v1alpha1.NatsJetStreamChannel) pkgreconciler.Event {
+	logger := logging.FromContext(ctx)
+	logger.Debugw("ReconcileKind for channel", zap.String("channel", nc.Name))
+
+	if err := r.reconcileStream(ctx, nc); err != nil {
+		logger.Errorw("failed to reconcile stream", zap.Error(err))
+		return err
+	}
+
+	return r.syncChannel(ctx, nc, true)
+}
+
+// ObserveKind is invoked when a NatsJetStreamChannel requires reconciliation but the dispatcher is not the leader.
+// This will wait until the channel has been marked "ready" by the leader, then subscribe to the JSM stream to forward
+// to Knative subscriptions; this enables us to scale dispatchers horizontally to cope with large message volumes.
+// The only requirement to allow this to work is the use of Queue Subscribers.
+func (r *Reconciler) ObserveKind(ctx context.Context, nc *v1alpha1.NatsJetStreamChannel) pkgreconciler.Event {
+	logger := logging.FromContext(ctx)
+	logger.Debugw("ObserveKind for channel", zap.String("channel", nc.Name))
+
+	return r.syncChannel(ctx, nc, false)
+}
+
+func (r *Reconciler) FinalizeKind(ctx context.Context, nc *v1alpha1.NatsJetStreamChannel) pkgreconciler.Event {
+	panic("implement me")
+}
+
+// reconcileStream ensures that the stream exists with the correct configuration.
+func (r *Reconciler) reconcileStream(ctx context.Context, nc *v1alpha1.NatsJetStreamChannel) pkgreconciler.Event {
+	logger := logging.FromContext(ctx)
+
+	streamName := r.streamNameFunc(nc)
+	primarySubject := r.dispatcher.subjectFunc(nc.Namespace, nc.Name)
+
+	existing, err := r.js.StreamInfo(streamName)
+	if err != nil && err != nats.ErrStreamNotFound {
+		logger.Errorw("failed to check current stream info")
+		return err
+	}
+
+	streamConfig := buildStreamConfig(streamName, primarySubject, nc.Spec.Stream.Config)
+	isCreating := existing == nil
+
+	// AddStream is idempotent if the config is identical to that on the server
+	info, err := r.js.AddStream(streamConfig)
+	if err != nil {
+		logger.Errorw("failed to add stream")
+		controller.GetEventRecorder(ctx).Event(nc, corev1.EventTypeWarning, ReasonJetstreamStreamFailed, err.Error())
+		nc.Status.MarkStreamFailed("DispatcherCreateStreamFailed", "Failed to create JetStream stream")
+		return err
+	}
+
+	if isCreating {
+		logger.Infow("jetstream stream created", zap.String("stream_name", info.Config.Name))
+		controller.GetEventRecorder(ctx).Event(nc, corev1.EventTypeNormal, ReasonJetstreamStreamCreated, "Dispatcher deployment created")
+	}
+
+	nc.Status.MarkStreamTrue()
+
+	return nil
+}
+
+func (r *Reconciler) syncChannel(ctx context.Context, nc *v1alpha1.NatsJetStreamChannel, isLeader bool) pkgreconciler.Event {
+	logger := logging.FromContext(ctx).With(zap.Bool("is_leader", isLeader))
+
+	if !nc.Status.IsReady() {
+		logger.Debugw("NatsJetStreamChannel still not ready, short-circuiting the reconciler", zap.String("channel", nc.Name))
+		return nil
+	}
```

Comment:
Thoughts... If we're not syncing the dispatcher with hostnames to listen for until after the channel is ready, then this means a sender could be sending messages to a channel before the dispatcher is ready to handle them.

If we're in a single replica dispatcher deployment, this can't happen since the channel won't be marked as ready until after the stream and consumer are created. So in this scenario there is no problem

This breaks down when there are multiple dispatcher replicas. The "leader" will mark the channel as ready, but other non-leader dispatcher pods may not have registered the hostnames yet. Since in k8s eyes this dispatcher pod is "ready", requests may be routed to it, but the receiver won't know how to handle it.

Not sure how to handle this:

1. Simple solution, but not ideal... Only allow single replicas
2. If a request is received on a host not registered, go and query the channel and manually call `ObserveKind` within the scope of the request. This will be slower, but at least it's handled, then the proper reconcile (which is probably only milliseconds away) will be a no-op.

Anybody any better ideas?